### ISSUE TITLE
Possible bug in  build_SparseArrayTable

### ISSUE THREAD

**Issue by mabobr** [rsyslog/rsyslog#4490]
Function _build_SparseArrayTable()_ has this code:

`pThis->table.sprsArr->entries[i].key = (uint32_t) json_object_get_int(jindex);`

when value of jindex is >  2147483647 then **value 2147483647 is always returned/used**  which breaks searches for values above this limit.
Also function _build_ArrayTable()_ uses _json_object_get_int()_ and must be checked also.
Perhaps _int64_t json_object_get_int64()_ may help somehow.

Related to issue: lookup on sparsearray - wrong result #3394

**Comment by rgerhards** [comment]
@janmejay do you happen to have a couple of minutes to comment on this?

**Comment by janmejay** [comment]
@mabobr Yep, that certainly seems to be a problem. Since we use libfastjson (our own fork), we should probably create another function `json_object_get_uint32` and use that here and in other such call-sites.

**Comment by janmejay** [comment]
@mabobr @rgerhards does [this](https://github.com/rsyslog/libfastjson/pull/162) look ok?

**Comment by janmejay** [comment]
@mabobr @rgerhards actual [bugfix](https://github.com/rsyslog/rsyslog/pull/4502), uses the function the former PR adds.

**Comment by mabobr** [comment]
Hi, this should work for you example:

# how to test:
# install current rsyslog
# put this file into /etc/rsyslog.d
# /usr/sbin/rsyslogd -N1 && sudo systemctl restart rsyslog
# in terminal one run: nc -u -l 51514
# in terminal two run (beware on quotes, ticks):  echo
'{"time":"2020-02-02T18:00:00", "mode":"imuxsock", "host":"venus", "msg":"
{\"name\":\"bill\", \"address\":\"123 Anywhere\", \"City\":\"Cleveland\",
\"Stat
e\":\"Ohio\", \"zip\":\"12345\"}"}' >/dev/udp/127.0.0.1/2514
# check result in termnal 1

module(load="imudp")
template (name="t_parsed_to_json" type="string" string="%$!all-json%\n")
ruleset(
          name="nestedjson"
          queue.type="LinkedList"
){
 unset $!;
 set $.ret = parse_json( $rawmsg, "\$!");
 if $.ret != 0 then { set $!error = "1 - bad json on input"; }
 if strlen($!msg) == 0 then { set $!error = $!error & "2 - msg not found";
}
 set $.ret = parse_json( ltrim($!msg), "\$.lev2");
 if $.ret != 0 then { set $!error = $!error & "3 - inner msg is not json";
}
 unset $!msg;
 set $! = $.lev2;
 action( type="omfwd"  Target="127.0.0.1" Port="51514" Protocol="udp"
template="t_parsed_to_json")
 stop
}
input(type="imudp" port="2514" ruleset="nestedjson" )

**Comment by rgerhards** [comment]
closed via #4507
